File: api/Helpers/ApiResponse.cs
using Microsoft.AspNetCore.Mvc;

namespace ProjectManagementAPI.Helpers;

// ═══════════════════════════════════════════════════════════════════════════
//  ApiResponse — envelope เดียวกันทุก endpoint
// ═══════════════════════════════════════════════════════════════════════════
public record ApiResponse<T>
{
    public bool Success { get; init; }
    public T? Data { get; init; }
    public string? Message { get; init; }
    public string? Code { get; init; }

    public static ApiResponse<T> Ok(T data, string? message = null)
        => new() { Success = true, Data = data, Message = message };

    public static ApiResponse<T> Fail(string message, string? code = null)
        => new() { Success = false, Message = message, Code = code };
}

/// <summary>envelope สำหรับ endpoint ที่ไม่มี data กลับ</summary>
public record ApiResponse : ApiResponse<object>
{
    public static ApiResponse Ok(string? message = null)
        => new() { Success = true, Message = message };
}

public record PagedResult<T>(IReadOnlyList<T> Items, int Total, int Page, int PageSize)
{
    public bool HasMore => Page * PageSize < Total;
}

// ═══════════════════════════════════════════════════════════════════════════
//  Result → HTTP mapping — "ที่เดียว" ตามที่ PLAN.md กำหนด
//
//  controller ไม่ต้องรู้ว่า ErrorKind ไหนแปลงเป็น status อะไร แค่เรียก
//  ToActionResult() ทำให้ status code ทั้งระบบสอดคล้องกันโดยไม่ต้องพึ่งวินัย
//
//  ⚠️ NotFound และ Forbidden ที่เกิดจาก "ไม่มีสิทธิ์เห็น" ต้องตอบ 404 ทั้งคู่
//     403 บอกผู้ใช้ว่า resource นี้มีอยู่จริง = leak การมีอยู่ข้าม tenant
// ═══════════════════════════════════════════════════════════════════════════
public static class ResultExtensions
{
    public static IActionResult ToActionResult<T>(this Result<T> result)
        => result.IsSuccess
            ? new OkObjectResult(ApiResponse<T>.Ok(result.Value))
            : ToErrorResult(result.Error);

    public static IActionResult ToActionResult(this Result result)
        => result.IsSuccess
            ? new OkObjectResult(ApiResponse.Ok())
            : ToErrorResult(result.Error);

    public static IActionResult ToCreatedResult<T>(this Result<T> result, string location)
        => result.IsSuccess
            ? new CreatedResult(location, ApiResponse<T>.Ok(result.Value))
            : ToErrorResult(result.Error);

    private static IActionResult ToErrorResult(Error error)
    {
        var statusCode = error.Kind switch
        {
            ErrorKind.NotFound => StatusCodes.Status404NotFound,
            ErrorKind.Validation => StatusCodes.Status400BadRequest,
            ErrorKind.Unauthorized => StatusCodes.Status401Unauthorized,
            ErrorKind.Forbidden => StatusCodes.Status403Forbidden,
            ErrorKind.Conflict => StatusCodes.Status409Conflict,
            ErrorKind.Unavailable => StatusCodes.Status503ServiceUnavailable,
            _ => StatusCodes.Status500InternalServerError
        };

        return new ObjectResult(ApiResponse<object>.Fail(error.Message, error.Code))
        {
            StatusCode = statusCode
        };
    }
}

File: server/src/common/api-response.ts
import { HttpException, HttpStatus } from '@nestjs/common';

import type { AppError, ErrorKind, Result } from './result.js';

// ═══════════════════════════════════════════════════════════════════════════
//  ApiResponse — envelope เดียวกันทุก endpoint
//
//  ⚠️ รูปนี้เป็นสัญญากับ web/ อยู่แล้ว (ดู ApiEnvelope ใน web/src/lib/apiClient.ts)
//     เปลี่ยนชื่อ field ที่นี่ = แก้ฝั่งเว็บด้วยในคอมมิตเดียวกัน
// ═══════════════════════════════════════════════════════════════════════════
export interface ApiEnvelope<T> {
  success: boolean;
  data?: T;
  message?: string;
  code?: string;
}

export const envelopeOk = <T>(data: T, message?: string): ApiEnvelope<T> =>
  message === undefined ? { success: true, data } : { success: true, data, message };

export const envelopeFail = (message: string, code?: string): ApiEnvelope<never> =>
  code === undefined ? { success: false, message } : { success: false, message, code };

// ═══════════════════════════════════════════════════════════════════════════
//  Result → HTTP — "ที่เดียว"
//
//  controller ไม่ต้องรู้ว่า ErrorKind ไหนแปลงเป็น status อะไร แค่เรียก unwrap()
//  ทำให้ status code ทั้งระบบสอดคล้องกันโดยไม่ต้องพึ่งวินัย
//
//  ⚠️ "ไม่พบ" กับ "ไม่มีสิทธิ์เห็น" ต้องตอบ 404 ทั้งคู่
//     403 บอกผู้ใช้ว่า resource นี้มีอยู่จริง = leak การมีอยู่ข้าม tenant
//     (ที่ที่ตั้งใจตอบ 403 คือกรณีที่ผู้เรียก "เห็น" ของนั้นได้อยู่แล้ว เช่น
//      เป็นสมาชิก workspace แต่ role ต่ำเกินไป)
// ═══════════════════════════════════════════════════════════════════════════
const STATUS: Record<ErrorKind, HttpStatus> = {
  not_found: HttpStatus.NOT_FOUND,
  validation: HttpStatus.BAD_REQUEST,
  unauthorized: HttpStatus.UNAUTHORIZED,
  forbidden: HttpStatus.FORBIDDEN,
  conflict: HttpStatus.CONFLICT,
  unavailable: HttpStatus.SERVICE_UNAVAILABLE,
};

/** exception ที่ถือ envelope ไว้แล้ว — ApiExceptionFilter ส่งต่อตรง ๆ */
export class ApiException extends HttpException {
  constructor(error: AppError) {
    super(envelopeFail(error.message, error.code), STATUS[error.kind]);
  }
}

export const toApiException = (error: AppError): ApiException => new ApiException(error);

/**
 * แกะ Result ที่สำเร็จออกมา หรือ throw ให้ Nest ตอบ error ที่ถูกต้อง
 *
 * ⚠️ เรียกที่ controller เท่านั้น — service ที่เรียก service อื่นควรส่ง Result
 *    ต่อไปตามสายด้วย propagate() ไม่ใช่ throw กลางทาง ไม่งั้นเหตุผลข้อ 1 ของ
 *    การมี Result (อ่าน signature แล้วรู้ว่าล้มเหลวได้แบบไหน) หายไปทันที
 */
export function unwrap<T>(result: Result<T>): T {
  if (!result.ok) throw new ApiException(result.error);
  return result.value;
}
